### PR TITLE
First step to refactoring S3 reading logic

### PR DESCRIPTION
**Pull request by janeyx99**
This is an initial attempt in refactoring and consolidating our S3 read logic for print_test_stats.py, test_history.py, and run_test.py. This way, boto3 and botocore do not need to be imported in various places throughout the code base, and duplicated logic (such as the many type definitions) can exist in one place: `tools/stat_utils/s3_stat_parser.py`. @walterddr contributed to this PR by moving print_test_stats.py to the tools folder and the corresponding tests a subfolder within tools. 

**NOTE: this removes those tests from CI as the new `tools/test/test_stats.py` is not in the test/ directory as the other tests in TESTS in run_test.py.**

Test plan: 
This refactoring change should not break anything, so running the files as before should work as they did previously.
To make sure that print_test_stats.py still functions: run `python tools/test/test_stats.py` and make sure all tests pass.
To make sure that test_history.py works, run the example commands from `tools/test_history.py --help` and check that their output matches that shown. Note that the script will continue printing for a while, so don't be alarmed.

Some next steps:
- Actually coming up with similarities among the three current use cases and further refactoring/consolidating of functions (e.g., combining simplify and get_cases)
- Moving more parsing logic to s3_stat_parser.py to have better abstraction between our files
- Adding tests for s3_stat_parser.py when there is more functionality in it

### CODE REVIEW

File: test/run_test.py [hunk 1]
```diff
@@ -443,6 +438,7 @@ def calculate_shards(num_shards: int, tests: List[str], job_times: Dict[str, Tup
 
 
 def pull_job_times_from_S3() -> Dict[str, Tuple[float, int]]:
+    print("HAHAHAHAHA")
```
Comment (janeyx99) [inline test/run_test.py:441]: lol i have to remove this

**Review by walterddr — COMMENTED** [review]
looks good (not suppose to accept since I am part of contributor)
but you might need to manual rebase once #53942 is landed, it seems none-trivial

File: test/run_test.py [hunk 2]
```diff
@@ -16,17 +16,18 @@
...
+
 import torch.distributed as dist
 from typing import Dict, Optional, Tuple, List, Any
 
 try:
-    import boto3  # type: ignore[import]
-    import botocore  # type: ignore[import]
-    import botocore.exceptions  # type: ignore[import]
-    HAVE_BOTO3 = True
+    sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), ".."))
```
Comment (walterddr) [inline test/run_test.py:24]:
```suggestion
    sys.path.append(os.path.join(os.path.dirname(os.path.abspath(__file__)), ".."))
```

**Comment by samestep** [comment]
This resolves #52866, right?

**Comment by janeyx99** [comment]
> This resolves #52866, right?

A part of it. The tests are currently not on CI :/

**Review by samestep — APPROVED** [review]
Looks great, thanks so much for doing this! Might be better to have a more detailed test plan though (since `tools/test_history.py` never had automated tests, and the tests this PR moves to `tools/test` are now no longer run in CI)

File: .circleci/config.yml [hunk 3]
```diff
@@ -636,7 +636,8 @@ jobs:
           export CIRCLE_JOB="$CIRCLE_JOB"
           export CIRCLE_WORKFLOW_ID="$CIRCLE_WORKFLOW_ID"
           cd workspace
-          python torch/testing/_internal/print_test_stats.py --upload-to-s3 --compare-with-s3 test
+          export PYTHONPATH="\${PWD}"
```
[the remark this answers is not in the document]
Comment (janeyx99) [reply .circleci/config.yml:639]: This is so it'll evaluate it in the Docker, and not right away. It escapes the $

**Pull request closed without merging**